[Bug] - sourcecode analyzer unit tests do not run offline

### Description
The unit tests for `pypi_sourcecode_analyzer.py` when no network connection is available will cause them to fail. The underlying reason for this is due to the network connections that are made by the subprocess Semgrep.

### Steps to Reproduce
1. **Step 1**: Run `pytest tests/malware_analyzer/pypi/test_pypi_sourcecode_analyzer.py` with no network connection.

### Expected Behavior
All tests should pass.

### Actual Behavior
The following three tests fail:
```
tests/malware_analyzer/pypi/test_pypi_sourcecode_analyzer.py::test_custom_rules FAILED
tests/malware_analyzer/pypi/test_pypi_sourcecode_analyzer.py::test_disabling_rulesets[test_disable_custom_ruleset] FAILED
tests/malware_analyzer/pypi/test_pypi_sourcecode_analyzer.py::test_disabling_rules FAILED
```

### Debug Information
The following errors are raised and cause the tests to fail:
```
_________________________________________________________________________________________________ test_custom_rules _________________________________________________________________________________________________
Traceback (most recent call last):
  File "/home/carl_flottmann/macaron/src/macaron/malware_analyzer/pypi_heuristics/sourcecode/pypi_sourcecode_analyzer.py", line 116, in _load_defaults
    process = subprocess.run(semgrep_commands, check=True, capture_output=True)  # nosec B603
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['semgrep', 'scan', '--validate', '--oss-only', '--config', '/home/carl_flottmann/macaron/tests/malware_analyzer/pypi/resources/custom_sample.yaml']' returned non-zero exit status 2.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/_pytest/runner.py", line 344, in from_call
    result: TResult | None = func()
                             ^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/_pytest/runner.py", line 246, in <lambda>
    lambda: runtest_hook(item=item, **kwds), when=when, reraise=reraise
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_hooks.py", line 512, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 167, in _multicall
    raise exception
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 139, in _multicall
    teardown.throw(exception)
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/_pytest/logging.py", line 850, in pytest_runtest_call
    yield
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 139, in _multicall
    teardown.throw(exception)
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 53, in run_old_style_hookwrapper
    return result.get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_result.py", line 103, in get_result
    raise exc.with_traceback(tb)
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 38, in run_old_style_hookwrapper
    res = yield
          ^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 139, in _multicall
    teardown.throw(exception)
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/_pytest/capture.py", line 900, in pytest_runtest_call
    return (yield)
            ^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 139, in _multicall
    teardown.throw(exception)
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 53, in run_old_style_hookwrapper
    return result.get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_result.py", line 103, in get_result
    raise exc.with_traceback(tb)
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 38, in run_old_style_hookwrapper
    res = yield
          ^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 139, in _multicall
    teardown.throw(exception)
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/_pytest/skipping.py", line 263, in pytest_runtest_call
    return (yield)
            ^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 121, in _multicall
    res = hook_impl.function(*args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/_pytest/runner.py", line 178, in pytest_runtest_call
    item.runtest()
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/_pytest/python.py", line 1670, in runtest
    self.ihook.pytest_pyfunc_call(pyfuncitem=self)
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_hooks.py", line 512, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 167, in _multicall
    raise exception
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 121, in _multicall
    res = hook_impl.function(*args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/_pytest/python.py", line 156, in pytest_pyfunc_call
    result = testfunction(**testargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/unittest/mock.py", line 1378, in patched
    return func(*newargs, **newkeywargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/tests/malware_analyzer/pypi/test_pypi_sourcecode_analyzer.py", line 163, in test_custom_rules
    analyzer = PyPISourcecodeAnalyzer(resources_path=RESOURCES_PATH)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/src/macaron/malware_analyzer/pypi_heuristics/sourcecode/pypi_sourcecode_analyzer.py", line 62, in __init__
    self.default_rule_path, self.custom_rule_path, self.disabled_rule_ids = self._load_defaults(resources_path)
                                                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/src/macaron/malware_analyzer/pypi_heuristics/sourcecode/pypi_sourcecode_analyzer.py", line 123, in _load_defaults
    raise ConfigurationError(error_msg) from semgrep_error
macaron.errors.ConfigurationError: Unable to run semgrep validation on /home/carl_flottmann/macaron/tests/malware_analyzer/pypi/resources/custom_sample.yaml with arguments ['semgrep', 'scan', '--validate', '--oss-only', '--config', '/home/carl_flottmann/macaron/tests/malware_analyzer/pypi/resources/custom_sample.yaml']: Command '['semgrep', 'scan', '--validate', '--oss-only', '--config', '/home/carl_flottmann/macaron/tests/malware_analyzer/pypi/resources/custom_sample.yaml']' returned non-zero exit status 2..
_______________________________________________________________________________ test_disabling_rulesets[test_disable_custom_ruleset] ________________________________________________________________________________
Traceback (most recent call last):
  File "/home/carl_flottmann/macaron/src/macaron/malware_analyzer/pypi_heuristics/sourcecode/pypi_sourcecode_analyzer.py", line 116, in _load_defaults
    process = subprocess.run(semgrep_commands, check=True, capture_output=True)  # nosec B603
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['semgrep', 'scan', '--validate', '--oss-only', '--config', '/home/carl_flottmann/macaron/tests/malware_analyzer/pypi/resources']' returned non-zero exit status 2.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/_pytest/runner.py", line 344, in from_call
    result: TResult | None = func()
                             ^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/_pytest/runner.py", line 246, in <lambda>
    lambda: runtest_hook(item=item, **kwds), when=when, reraise=reraise
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_hooks.py", line 512, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 167, in _multicall
    raise exception
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 139, in _multicall
    teardown.throw(exception)
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/_pytest/logging.py", line 850, in pytest_runtest_call
    yield
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 139, in _multicall
    teardown.throw(exception)
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 53, in run_old_style_hookwrapper
    return result.get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_result.py", line 103, in get_result
    raise exc.with_traceback(tb)
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 38, in run_old_style_hookwrapper
    res = yield
          ^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 139, in _multicall
    teardown.throw(exception)
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/_pytest/capture.py", line 900, in pytest_runtest_call
    return (yield)
            ^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 139, in _multicall
    teardown.throw(exception)
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 53, in run_old_style_hookwrapper
    return result.get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_result.py", line 103, in get_result
    raise exc.with_traceback(tb)
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 38, in run_old_style_hookwrapper
    res = yield
          ^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 139, in _multicall
    teardown.throw(exception)
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/_pytest/skipping.py", line 263, in pytest_runtest_call
    return (yield)
            ^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 121, in _multicall
    res = hook_impl.function(*args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/_pytest/runner.py", line 178, in pytest_runtest_call
    item.runtest()
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/_pytest/python.py", line 1670, in runtest
    self.ihook.pytest_pyfunc_call(pyfuncitem=self)
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_hooks.py", line 512, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 167, in _multicall
    raise exception
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 121, in _multicall
    res = hook_impl.function(*args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/_pytest/python.py", line 156, in pytest_pyfunc_call
    result = testfunction(**testargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/unittest/mock.py", line 1378, in patched
    return func(*newargs, **newkeywargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/tests/malware_analyzer/pypi/test_pypi_sourcecode_analyzer.py", line 222, in test_disabling_rulesets
    analyzer = PyPISourcecodeAnalyzer(resources_path=RESOURCES_PATH)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/src/macaron/malware_analyzer/pypi_heuristics/sourcecode/pypi_sourcecode_analyzer.py", line 62, in __init__
    self.default_rule_path, self.custom_rule_path, self.disabled_rule_ids = self._load_defaults(resources_path)
                                                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/src/macaron/malware_analyzer/pypi_heuristics/sourcecode/pypi_sourcecode_analyzer.py", line 123, in _load_defaults
    raise ConfigurationError(error_msg) from semgrep_error
macaron.errors.ConfigurationError: Unable to run semgrep validation on /home/carl_flottmann/macaron/tests/malware_analyzer/pypi/resources with arguments ['semgrep', 'scan', '--validate', '--oss-only', '--config', '/home/carl_flottmann/macaron/tests/malware_analyzer/pypi/resources']: Command '['semgrep', 'scan', '--validate', '--oss-only', '--config', '/home/carl_flottmann/macaron/tests/malware_analyzer/pypi/resources']' returned non-zero exit status 2..
_______________________________________________________________________________________________ test_disabling_rules ________________________________________________________________________________________________
Traceback (most recent call last):
  File "/home/carl_flottmann/macaron/src/macaron/malware_analyzer/pypi_heuristics/sourcecode/pypi_sourcecode_analyzer.py", line 116, in _load_defaults
    process = subprocess.run(semgrep_commands, check=True, capture_output=True)  # nosec B603
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['semgrep', 'scan', '--validate', '--oss-only', '--config', '/home/carl_flottmann/macaron/tests/malware_analyzer/pypi/resources/custom_sample.yaml']' returned non-zero exit status 2.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/_pytest/runner.py", line 344, in from_call
    result: TResult | None = func()
                             ^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/_pytest/runner.py", line 246, in <lambda>
    lambda: runtest_hook(item=item, **kwds), when=when, reraise=reraise
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_hooks.py", line 512, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 167, in _multicall
    raise exception
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 139, in _multicall
    teardown.throw(exception)
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/_pytest/logging.py", line 850, in pytest_runtest_call
    yield
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 139, in _multicall
    teardown.throw(exception)
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 53, in run_old_style_hookwrapper
    return result.get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_result.py", line 103, in get_result
    raise exc.with_traceback(tb)
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 38, in run_old_style_hookwrapper
    res = yield
          ^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 139, in _multicall
    teardown.throw(exception)
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/_pytest/capture.py", line 900, in pytest_runtest_call
    return (yield)
            ^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 139, in _multicall
    teardown.throw(exception)
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 53, in run_old_style_hookwrapper
    return result.get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_result.py", line 103, in get_result
    raise exc.with_traceback(tb)
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 38, in run_old_style_hookwrapper
    res = yield
          ^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 139, in _multicall
    teardown.throw(exception)
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/_pytest/skipping.py", line 263, in pytest_runtest_call
    return (yield)
            ^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 121, in _multicall
    res = hook_impl.function(*args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/_pytest/runner.py", line 178, in pytest_runtest_call
    item.runtest()
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/_pytest/python.py", line 1670, in runtest
    self.ihook.pytest_pyfunc_call(pyfuncitem=self)
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_hooks.py", line 512, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 167, in _multicall
    raise exception
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 121, in _multicall
    res = hook_impl.function(*args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/.venv/lib/python3.11/site-packages/_pytest/python.py", line 156, in pytest_pyfunc_call
    result = testfunction(**testargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/unittest/mock.py", line 1378, in patched
    return func(*newargs, **newkeywargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/tests/malware_analyzer/pypi/test_pypi_sourcecode_analyzer.py", line 278, in test_disabling_rules
    analyzer = PyPISourcecodeAnalyzer(resources_path=RESOURCES_PATH)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/src/macaron/malware_analyzer/pypi_heuristics/sourcecode/pypi_sourcecode_analyzer.py", line 62, in __init__
    self.default_rule_path, self.custom_rule_path, self.disabled_rule_ids = self._load_defaults(resources_path)
                                                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/carl_flottmann/macaron/src/macaron/malware_analyzer/pypi_heuristics/sourcecode/pypi_sourcecode_analyzer.py", line 123, in _load_defaults
    raise ConfigurationError(error_msg) from semgrep_error
macaron.errors.ConfigurationError: Unable to run semgrep validation on /home/carl_flottmann/macaron/tests/malware_analyzer/pypi/resources/custom_sample.yaml with arguments ['semgrep', 'scan', '--validate', '--oss-only', '--config', '/home/carl_flottmann/macaron/tests/malware_analyzer/pypi/resources/custom_sample.yaml']: Command '['semgrep', 'scan', '--validate', '--oss-only', '--config', '/home/carl_flottmann/macaron/tests/malware_analyzer/pypi/resources/custom_sample.yaml']' returned non-zero exit status 2..
```

```
============================================================================================== short test summary info ==============================================================================================
FAILED tests/malware_analyzer/pypi/test_pypi_sourcecode_analyzer.py::test_custom_rules - macaron.errors.ConfigurationError: Unable to run semgrep validation on /home/carl_flottmann/macaron/tests/malware_analyzer/pypi/resources/custom_sample.yaml with arguments ['semgrep', 'scan', '--validate', '--oss-only', '--config', '/home/carl_flottmann/macaron/tests/malware_analyzer/pypi/resources/custom_sample.yaml']: Command '['semgrep', 'scan', '--validate', '--oss-only', '--config', '/home/carl_flottmann/macaron/tests/malware_analyzer/pypi/resources/custom_sample.yaml']' returned non-zero exit status 2..
FAILED tests/malware_analyzer/pypi/test_pypi_sourcecode_analyzer.py::test_disabling_rulesets[test_disable_custom_ruleset] - macaron.errors.ConfigurationError: Unable to run semgrep validation on /home/carl_flottmann/macaron/tests/malware_analyzer/pypi/resources with arguments ['semgrep', 'scan', '--validate', '--oss-only', '--config', '/home/carl_flottmann/macaron/tests/malware_analyzer/pypi/resources']: Command '['semgrep', 'scan', '--validate', '--oss-only', '--config', '/home/carl_flottmann/macaron/tests/malware_analyzer/pypi/resources']' returned non-zero exit status 2..
FAILED tests/malware_analyzer/pypi/test_pypi_sourcecode_analyzer.py::test_disabling_rules - macaron.errors.ConfigurationError: Unable to run semgrep validation on /home/carl_flottmann/macaron/tests/malware_analyzer/pypi/resources/custom_sample.yaml with arguments ['semgrep', 'scan', '--validate', '--oss-only', '--config', '/home/carl_flottmann/macaron/tests/malware_analyzer/pypi/resources/custom_sample.yaml']: Command '['semgrep', 'scan', '--validate', '--oss-only', '--config', '/home/carl_flottmann/macaron/tests/malware_analyzer/pypi/resources/custom_sample.yaml']' returned non-zero exit status 2..
===================================================================================== 3 failed, 10 passed in 634.53s (0:10:34) ======================================================================================
```


### Environment Information
Operating System: 24.04.1 LTS (Noble Numbat) on WSL version 2.5.9.0

CPU architecture information: x86-64

Bash Version: 5.2.21(1)-release (x86_64-pc-linux-gnu)

Python virtual environment, version: 3.11.10

--- a/src/macaron/malware_analyzer/pypi_heuristics/sourcecode/pypi_sourcecode_analyzer.py
+++ b/src/macaron/malware_analyzer/pypi_heuristics/sourcecode/pypi_sourcecode_analyzer.py
@@ -118,7 +118,17 @@ class PyPISourcecodeAnalyzer(BaseHeuristicAnalyzer):
                 logger.debug(error_msg)
                 raise ConfigurationError(error_msg)
 
-            semgrep_commands: list[str] = ["semgrep", "scan", "--validate", "--oss-only", "--config", custom_rule_path]
+            semgrep_commands: list[str] = [
+                "semgrep",
+                "scan",
+                "--metrics",
+                "off",
+                "--disable-version-check",
+                "--validate",
+                "--oss-only",
+                "--config",
+                custom_rule_path,
+            ]
             try:
                 process = subprocess.run(semgrep_commands, check=True, capture_output=True)  # nosec B603
             except (subprocess.CalledProcessError, subprocess.TimeoutExpired) as semgrep_error:
@@ -248,7 +258,15 @@ class PyPISourcecodeAnalyzer(BaseHeuristicAnalyzer):
         # Here, we disable 'nosemgrep' ignoring so that this is not an evasion method of our scan (i.e. malware includes
         # 'nosemgrep' comments to prevent our scan detecting those code lines). Read more about the 'nosemgrep' feature
         # here: https://semgrep.dev/docs/ignoring-files-folders-code
-        semgrep_commands: list[str] = ["semgrep", "scan", "--oss-only", "--disable-nosem"]
+        semgrep_commands: list[str] = [
+            "semgrep",
+            "scan",
+            "--metrics",
+            "off",
+            "--disable-version-check",
+            "--oss-only",
+            "--disable-nosem",
+        ]
         result: HeuristicResult = HeuristicResult.PASS
 
         source_code_path = pypi_package_json.package_sourcecode_path

--- a/src/macaron/malware_analyzer/pypi_heuristics/sourcecode/pypi_sourcecode_analyzer.py
+++ b/src/macaron/malware_analyzer/pypi_heuristics/sourcecode/pypi_sourcecode_analyzer.py
@@ -143,7 +143,7 @@ class PyPISourcecodeAnalyzer(BaseHeuristicAnalyzer):
                     # Only a warning is used here, so that if running offline, the analysis can continue. Erroneous Semgrep files
                     # will be picked up at analysis time in this case.
                     warning_msg = (
-                        f"Running semgrep validation on {custom_rule_path} was with arguments"
+                        f"Running semgrep validation on {custom_rule_path} with argument(s)"
                         f" {process.args} "
                         f" was not successful: {process.returncode}."
                         " These custom rule(s) may not run successfully."
@@ -152,7 +152,7 @@ class PyPISourcecodeAnalyzer(BaseHeuristicAnalyzer):
 
             except (subprocess.CalledProcessError, subprocess.TimeoutExpired) as semgrep_error:
                 warning_msg = (
-                    f"Unable to run semgrep validation on {custom_rule_path} with arguments "
+                    f"Unable to run semgrep validation on {custom_rule_path} with argument(s) "
                     f"{semgrep_commands}: {semgrep_error}."
                 )
                 logger.warning(warning_msg)
@@ -306,13 +306,13 @@ class PyPISourcecodeAnalyzer(BaseHeuristicAnalyzer):
                 process = subprocess.run(semgrep_commands, check=True, capture_output=True)  # nosec B603
             except (subprocess.CalledProcessError, subprocess.TimeoutExpired) as semgrep_error:
                 error_msg = (
-                    f"Unable to run semgrep on {source_code_path} with arguments {semgrep_commands}: {semgrep_error}"
+                    f"Unable to run semgrep on {source_code_path} with argument(s) {semgrep_commands}: {semgrep_error}"
                 )
                 logger.debug(error_msg)
                 raise HeuristicAnalyzerValueError(error_msg) from semgrep_error
 
             if process.returncode != 0:
-                error_msg = f"Error running semgrep on {source_code_path} with arguments" f" {process.args}"
+                error_msg = f"Error running semgrep on {source_code_path} with argument(s)" f" {process.args}"
                 logger.debug(error_msg)
                 raise HeuristicAnalyzerValueError(error_msg)
 

--- a/src/macaron/malware_analyzer/pypi_heuristics/sourcecode/pypi_sourcecode_analyzer.py
+++ b/src/macaron/malware_analyzer/pypi_heuristics/sourcecode/pypi_sourcecode_analyzer.py
@@ -118,6 +118,14 @@ class PyPISourcecodeAnalyzer(BaseHeuristicAnalyzer):
                 logger.debug(error_msg)
                 raise ConfigurationError(error_msg)
 
+            # Extra argument explanation:
+            # --metrics off is used to disable metric collection. This makes a network connection, which can
+            # impact running Semgrep offline.
+            # --disable-version-check is used to disable the network connection made to semgrep.dev to check
+            # if this is the latest version. This network connection can also impact running Semgrep offline.
+            # --oss-only is used to ensure only the open-source offering of Semgrep is run.
+            # Note, validation with --validate still currently makes a network connection to download linting
+            # rules, which cannot be turned off.
             semgrep_commands: list[str] = [
                 "semgrep",
                 "scan",
@@ -132,8 +140,13 @@ class PyPISourcecodeAnalyzer(BaseHeuristicAnalyzer):
             try:
                 process = subprocess.run(semgrep_commands, check=True, capture_output=True)  # nosec B603
                 if process.returncode != 0:
+                    # Only a warning is used here, so that if running offline, the analysis can continue. Erroneous Semgrep files
+                    # will be picked up at analysis time in this case.
                     warning_msg = (
-                        f"Error running semgrep validation on {custom_rule_path} with arguments" f" {process.args}."
+                        f"Running semgrep validation on {custom_rule_path} was with arguments"
+                        f" {process.args} "
+                        f" was not successful: {process.returncode}."
+                        " These custom rule(s) may not run successfully."
                     )
                     logger.warning(warning_msg)
 
@@ -255,9 +268,15 @@ class PyPISourcecodeAnalyzer(BaseHeuristicAnalyzer):
         analysis_result: dict = {}
         # Since we have to run them anyway, return disabled rule findings for debug information.
         disabled_results: dict = {}
-        # Here, we disable 'nosemgrep' ignoring so that this is not an evasion method of our scan (i.e. malware includes
-        # 'nosemgrep' comments to prevent our scan detecting those code lines). Read more about the 'nosemgrep' feature
-        # here: https://semgrep.dev/docs/ignoring-files-folders-code
+        # Extra argument explanation:
+        # --metrics off is used to disable metric collection. This makes a network connection, which can
+        # impact running Semgrep offline.
+        # --disable-version-check is used to disable the network connection made to semgrep.dev to check
+        # if this is the latest version. This network connection can also impact running Semgrep offline.
+        # --oss-only is used to ensure only the open-source offering of Semgrep is run.
+        # --disable-nosem is used to disable 'nosemgrep' ignoring so that this is not an evasion method of
+        # our scan (i.e. malware includes 'nosemgrep' comments to prevent our scan detecting those code lines).
+        # Read more about the 'nosemgrep' feature here: https://semgrep.dev/docs/ignoring-files-folders-code
         semgrep_commands: list[str] = [
             "semgrep",
             "scan",

--- a/src/macaron/malware_analyzer/pypi_heuristics/sourcecode/pypi_sourcecode_analyzer.py
+++ b/src/macaron/malware_analyzer/pypi_heuristics/sourcecode/pypi_sourcecode_analyzer.py
@@ -131,18 +131,18 @@ class PyPISourcecodeAnalyzer(BaseHeuristicAnalyzer):
             ]
             try:
                 process = subprocess.run(semgrep_commands, check=True, capture_output=True)  # nosec B603
+                if process.returncode != 0:
+                    warning_msg = (
+                        f"Error running semgrep validation on {custom_rule_path} with arguments" f" {process.args}."
+                    )
+                    logger.warning(warning_msg)
+
             except (subprocess.CalledProcessError, subprocess.TimeoutExpired) as semgrep_error:
-                error_msg = (
+                warning_msg = (
                     f"Unable to run semgrep validation on {custom_rule_path} with arguments "
                     f"{semgrep_commands}: {semgrep_error}."
                 )
-                logger.debug(error_msg)
-                raise ConfigurationError(error_msg) from semgrep_error
-
-            if process.returncode != 0:
-                error_msg = f"Error running semgrep validation on {custom_rule_path} with arguments" f" {process.args}."
-                logger.debug(error_msg)
-                raise ConfigurationError(error_msg)
+                logger.warning(warning_msg)
 
             logger.debug("Including custom ruleset from %s.", custom_rule_path)
 

--- a/tests/malware_analyzer/pypi/test_pypi_sourcecode_analyzer.py
+++ b/tests/malware_analyzer/pypi/test_pypi_sourcecode_analyzer.py
@@ -91,7 +91,7 @@ def test_nonexistent_rule_path(mock_defaults: MagicMock) -> None:
 
 
 @patch("macaron.malware_analyzer.pypi_heuristics.sourcecode.pypi_sourcecode_analyzer.defaults")
-def test_invalid_custom_rules(mock_defaults: MagicMock) -> None:
+def test_invalid_custom_rules(mock_defaults: MagicMock, pypi_package_json: MagicMock) -> None:
     """Test for when the provided file is not a valid semgrep rule, so error,"""
     # Use this file as an invalid semgrep rule as it is most definitely not a semgrep rule, and does exist.
     defaults = {
@@ -103,8 +103,14 @@ def test_invalid_custom_rules(mock_defaults: MagicMock) -> None:
     mock_defaults.has_section.side_effect = lambda section: section == "heuristic.pypi"
     mock_defaults.__getitem__.side_effect = lambda section: sub_section if section == "heuristic.pypi" else None
 
-    with pytest.raises(ConfigurationError):
-        _ = PyPISourcecodeAnalyzer(resources_path=RESOURCES_PATH)
+    analyzer = PyPISourcecodeAnalyzer(resources_path=RESOURCES_PATH)
+    pypi_package_json.package_sourcecode_path = os.path.join(
+        os.path.dirname(os.path.abspath(__file__)), "resources", "sourcecode_samples"
+    )
+
+    # Semgrep should fail to run when we launch analysis
+    with pytest.raises(HeuristicAnalyzerValueError):
+        _ = analyzer.analyze(pypi_package_json)
 
 
 @patch("macaron.malware_analyzer.pypi_heuristics.sourcecode.pypi_sourcecode_analyzer.defaults")

